{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "import dash\n",
    "from jupyter_dash import JupyterDash\n",
    "import dash_core_components as dcc\n",
    "import dash_bootstrap_components as dbc\n",
    "import dash_html_components as html\n",
    "import dash_table\n",
    "from dash.dependencies import Input, Output\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/paulli/anaconda3/envs/dash/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3441: DtypeWarning: Columns (39,41,42,43,44,49,50,51,53,54,58,59,61,62,70,71,75,76,78,79,83,84,85,86,87,88,89,90,91,92,93,94,95,96,98,99,100,101,102,103,104,105,106,107,108,110,111,112,115,117,118,119,122,123,124,125,126,127,128,131,132,133,134,135,136,137,138,139,140,141,142,143) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "filename = \"data/SraRunTable_wastewater.csv\"\n",
    "df_sra = pd.read_csv(filename)\n",
    "df_sra['collection_date'] = df_sra['collection_date'].str.split('/').str[0]\n",
    "df_sra['collection_date'] = pd.to_datetime(df_sra['collection_date'], errors='coerce', utc=True)\n",
    "df_sra['year'] = pd.DatetimeIndex(df_sra['collection_date']).year\n",
    "#df_sra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run,Assay Type,BioSample,Center Name,Consent,Experiment,Instrument,\n",
    "# LibraryLayout,LibrarySelection,LibrarySource,Organism,Platform,\n",
    "# ReleaseDate,Sample Name,SRA Study,DATASTORE filetype,DATASTORE provider,DATASTORE region,\n",
    "# AvgSpotLen,Bases,Bytes,BioProject,Library Name,\n",
    "# collection_date,geo_loc_name_country,geo_loc_name_country_continent,BioSampleModel,geo_loc_name,\n",
    "# lat_lon,isolation_source,env_broad_scale,env_local_scale,\n",
    "# Host,env_medium,ENA-FIRST-PUBLIC (run),ENA-LAST-UPDATE (run),geographic_location_(country_and/or_sea),\n",
    "# collected_by,sample_name,ww_population,ww_sample_duration,ww_sample_matrix,ww_sample_type,ww_surv_target_1_known_present,ww_surv_target_1,\n",
    "# INSDC_center_name,INSDC_first_public,INSDC_last_update,INSDC_status,env_biome,env_feature,env_material,INSDC_center_alias,\n",
    "# geographic_location_(latitude),geographic_location_(longitude),investigation_type,ena_first_public,\n",
    "# ena_last_update,purpose_of_ww_sampling,sequenced_by,isolate,ww_processing_protocol,ww_sample_site,project_name,Strain,External_Id,\n",
    "# Submitter_Id,replicate,SRA_accession,Title,ww_surv_jurisdiction,ww_surv_system_sample_id,\n",
    "# sewage_type,sequencing_method,environment_(biome),environment_(feature),environment_(material),\n",
    "# wastewater/sludge_environmental_package,Broker_name,samp_size,environmental_sample,\n",
    "# Is_the_sequenced_pathogen_host_associated?,isolation_source_non_host_associated,\n",
    "# Alias,ena_checklist,organism (run),collection_date (run),geo_loc_name (run),isolation_source (run),ww_population (run),ww_sample_duration (run),ww_sample_matrix (run),ww_sample_type (run),ww_surv_target_1 (run),\n",
    "# samp_mat_process,depth,samp_collect_device,id,name_of_the_sampling_site,rel_to_oxygen,misc_param,\n",
    "# description,source_material_ID,wastewater_type,host_subject_id,period_code,pH,site_code,host_sex,Host_Age,\n",
    "# host_body_product,replicate_code,Temp,reactor_type,geographic_location_(region_and_locality),\n",
    "# Tmp,isol_growth_condt,host_health_state,host_scientific_name,Bioreactor_Code,Sampling_week_designation,sample_type,\n",
    "# house,samp_vol_we_dna_ext,samp_store_temp,sample_name_2,sludge_retent_time,location,label,identity,sample_number,elev,\n",
    "# samp_store_dur,tot_nitro,BIOME,day,feature,material,wwtp,env_package,Antibiotic_amendment,\n",
    "# biochem_oxygen_dem,incubation_temp,Media_type,Timepoint,ref_biomaterial,nominal_length (run)\n",
    "\n",
    "# print(f'number of avil columns: {len(df_sra.columns)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimensions_dict = {\n",
    "    'assay_type': 'Assay Type', \n",
    "    'library_source': 'LibrarySource', \n",
    "    'platform': 'Platform', \n",
    "    'continent': 'geo_loc_name_country_continent',\n",
    "    'country': 'geo_loc_name_country'\n",
    "}\n",
    "dimensions_display = ['Assay Type', 'LibrarySource', 'Platform', 'geo_loc_name_country_continent']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_sra[dimensions_display].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SRA wastewater data description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/85/m0d0qjm51wj7gw6dgr_05lvw0009b7/T/ipykernel_34435/3058443914.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.\n",
      "  df_sra.describe(include='all')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Consent</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>...</th>\n",
       "      <th>wwtp</th>\n",
       "      <th>env_package</th>\n",
       "      <th>Antibiotic_amendment</th>\n",
       "      <th>biochem_oxygen_dem</th>\n",
       "      <th>incubation_temp</th>\n",
       "      <th>Media_type</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>ref_biomaterial</th>\n",
       "      <th>nominal_length (run)</th>\n",
       "      <th>year</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>...</td>\n",
       "      <td>249</td>\n",
       "      <td>244</td>\n",
       "      <td>234</td>\n",
       "      <td>234</td>\n",
       "      <td>234</td>\n",
       "      <td>234</td>\n",
       "      <td>234.0</td>\n",
       "      <td>190</td>\n",
       "      <td>225.000000</td>\n",
       "      <td>24890.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>25972</td>\n",
       "      <td>13</td>\n",
       "      <td>21749</td>\n",
       "      <td>385</td>\n",
       "      <td>1</td>\n",
       "      <td>25808</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>94</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>SRR1594353</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>SAMN20326102</td>\n",
       "      <td>BIOBOT ANALYTICS</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX3033573</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>Mallard Creek</td>\n",
       "      <td>MIGS/MIMS/MIMARKS.wastewater</td>\n",
       "      <td>no antibiotic</td>\n",
       "      <td>Not applicable</td>\n",
       "      <td>RT</td>\n",
       "      <td>LB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GAC and activate sludge</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>18381</td>\n",
       "      <td>1342</td>\n",
       "      <td>2213</td>\n",
       "      <td>25972</td>\n",
       "      <td>21</td>\n",
       "      <td>14489</td>\n",
       "      <td>21362</td>\n",
       "      <td>17023</td>\n",
       "      <td>16538</td>\n",
       "      <td>...</td>\n",
       "      <td>125</td>\n",
       "      <td>216</td>\n",
       "      <td>61</td>\n",
       "      <td>81</td>\n",
       "      <td>133</td>\n",
       "      <td>120</td>\n",
       "      <td>NaN</td>\n",
       "      <td>79</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>first</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>last</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>992.525778</td>\n",
       "      <td>2016.420571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2567.876966</td>\n",
       "      <td>3.349560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>2001.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>400.000000</td>\n",
       "      <td>2014.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>400.000000</td>\n",
       "      <td>2016.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>400.000000</td>\n",
       "      <td>2019.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19999.000000</td>\n",
       "      <td>2064.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows Ã— 148 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Run Assay Type     BioSample       Center Name Consent  \\\n",
       "count        25972      25972         25972             25972   25972   \n",
       "unique       25972         13         21749               385       1   \n",
       "top     SRR1594353   AMPLICON  SAMN20326102  BIOBOT ANALYTICS  public   \n",
       "freq             1      18381          1342              2213   25972   \n",
       "first          NaN        NaN           NaN               NaN     NaN   \n",
       "last           NaN        NaN           NaN               NaN     NaN   \n",
       "mean           NaN        NaN           NaN               NaN     NaN   \n",
       "std            NaN        NaN           NaN               NaN     NaN   \n",
       "min            NaN        NaN           NaN               NaN     NaN   \n",
       "25%            NaN        NaN           NaN               NaN     NaN   \n",
       "50%            NaN        NaN           NaN               NaN     NaN   \n",
       "75%            NaN        NaN           NaN               NaN     NaN   \n",
       "max            NaN        NaN           NaN               NaN     NaN   \n",
       "\n",
       "        Experiment      Instrument LibraryLayout LibrarySelection  \\\n",
       "count        25972           25972         25972            25972   \n",
       "unique       25808              35             2               12   \n",
       "top     SRX3033573  Illumina MiSeq        PAIRED              PCR   \n",
       "freq            21           14489         21362            17023   \n",
       "first          NaN             NaN           NaN              NaN   \n",
       "last           NaN             NaN           NaN              NaN   \n",
       "mean           NaN             NaN           NaN              NaN   \n",
       "std            NaN             NaN           NaN              NaN   \n",
       "min            NaN             NaN           NaN              NaN   \n",
       "25%            NaN             NaN           NaN              NaN   \n",
       "50%            NaN             NaN           NaN              NaN   \n",
       "75%            NaN             NaN           NaN              NaN   \n",
       "max            NaN             NaN           NaN              NaN   \n",
       "\n",
       "       LibrarySource  ...           wwtp                   env_package  \\\n",
       "count          25972  ...            249                           244   \n",
       "unique             6  ...              8                             3   \n",
       "top      METAGENOMIC  ...  Mallard Creek  MIGS/MIMS/MIMARKS.wastewater   \n",
       "freq           16538  ...            125                           216   \n",
       "first            NaN  ...            NaN                           NaN   \n",
       "last             NaN  ...            NaN                           NaN   \n",
       "mean             NaN  ...            NaN                           NaN   \n",
       "std              NaN  ...            NaN                           NaN   \n",
       "min              NaN  ...            NaN                           NaN   \n",
       "25%              NaN  ...            NaN                           NaN   \n",
       "50%              NaN  ...            NaN                           NaN   \n",
       "75%              NaN  ...            NaN                           NaN   \n",
       "max              NaN  ...            NaN                           NaN   \n",
       "\n",
       "       Antibiotic_amendment biochem_oxygen_dem incubation_temp Media_type  \\\n",
       "count                   234                234             234        234   \n",
       "unique                    5                 94               2          2   \n",
       "top           no antibiotic     Not applicable              RT         LB   \n",
       "freq                     61                 81             133        120   \n",
       "first                   NaN                NaN             NaN        NaN   \n",
       "last                    NaN                NaN             NaN        NaN   \n",
       "mean                    NaN                NaN             NaN        NaN   \n",
       "std                     NaN                NaN             NaN        NaN   \n",
       "min                     NaN                NaN             NaN        NaN   \n",
       "25%                     NaN                NaN             NaN        NaN   \n",
       "50%                     NaN                NaN             NaN        NaN   \n",
       "75%                     NaN                NaN             NaN        NaN   \n",
       "max                     NaN                NaN             NaN        NaN   \n",
       "\n",
       "       Timepoint          ref_biomaterial  nominal_length (run)          year  \n",
       "count      234.0                      190            225.000000  24890.000000  \n",
       "unique       NaN                       13                   NaN           NaN  \n",
       "top          NaN  GAC and activate sludge                   NaN           NaN  \n",
       "freq         NaN                       79                   NaN           NaN  \n",
       "first        NaN                      NaN                   NaN           NaN  \n",
       "last         NaN                      NaN                   NaN           NaN  \n",
       "mean         4.0                      NaN            992.525778   2016.420571  \n",
       "std          0.0                      NaN           2567.876966      3.349560  \n",
       "min          4.0                      NaN              0.300000   2001.000000  \n",
       "25%          4.0                      NaN            400.000000   2014.000000  \n",
       "50%          4.0                      NaN            400.000000   2016.000000  \n",
       "75%          4.0                      NaN            400.000000   2019.000000  \n",
       "max          4.0                      NaN          19999.000000   2064.000000  \n",
       "\n",
       "[13 rows x 148 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sra.describe(include='all')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SRA wastewater data description for plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>geo_loc_name_country</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>Platform</th>\n",
       "      <th>geo_loc_name_country_continent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972.0</td>\n",
       "      <td>2.597200e+04</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "      <td>25972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>25972</td>\n",
       "      <td>21749</td>\n",
       "      <td>902</td>\n",
       "      <td>731.0</td>\n",
       "      <td>2.450200e+04</td>\n",
       "      <td>385</td>\n",
       "      <td>110</td>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>SRR1594353</td>\n",
       "      <td>SAMN20326102</td>\n",
       "      <td>PRJNA746354</td>\n",
       "      <td>602.0</td>\n",
       "      <td>8.250000e+09</td>\n",
       "      <td>BIOBOT ANALYTICS</td>\n",
       "      <td>USA</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>North America</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>1342</td>\n",
       "      <td>2183</td>\n",
       "      <td>4989.0</td>\n",
       "      <td>9.000000e+01</td>\n",
       "      <td>2213</td>\n",
       "      <td>8465</td>\n",
       "      <td>18381</td>\n",
       "      <td>16538</td>\n",
       "      <td>24677</td>\n",
       "      <td>9832</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Run     BioSample   BioProject  AvgSpotLen         Bases  \\\n",
       "count        25972         25972        25972     25972.0  2.597200e+04   \n",
       "unique       25972         21749          902       731.0  2.450200e+04   \n",
       "top     SRR1594353  SAMN20326102  PRJNA746354       602.0  8.250000e+09   \n",
       "freq             1          1342         2183      4989.0  9.000000e+01   \n",
       "\n",
       "             Center Name geo_loc_name_country Assay Type LibrarySource  \\\n",
       "count              25972                25972      25972         25972   \n",
       "unique               385                  110         13             6   \n",
       "top     BIOBOT ANALYTICS                  USA   AMPLICON   METAGENOMIC   \n",
       "freq                2213                 8465      18381         16538   \n",
       "\n",
       "        Platform geo_loc_name_country_continent  \n",
       "count      25972                          25972  \n",
       "unique         7                              8  \n",
       "top     ILLUMINA                  North America  \n",
       "freq       24677                           9832  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df_sra[['Run', 'BioSample', 'BioProject', 'AvgSpotLen', 'Bases', 'Center Name', 'geo_loc_name_country']+dimensions_display]\n",
    "df = df.fillna('N/A')\n",
    "ddf = df\n",
    "ddf.describe(include='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fig_parallel_categories(df, dimensions, color_col):\n",
    "    ddf = df.copy()\n",
    "    \n",
    "    # set colors\n",
    "    c = list(ddf[color_col].unique())\n",
    "    colors = px.colors.qualitative.Dark24 + px.colors.qualitative.Light24\n",
    "    c_dict = dict(zip(c, colors[:len(c)]))\n",
    "    ddf['COLORS'] = ddf[color_col].map(c_dict)\n",
    "    \n",
    "    fig = px.parallel_categories(ddf, dimensions=dimensions, color='COLORS')\n",
    "    fig.update_layout(\n",
    "        margin=dict(l=20, r=20, t=20, b=20),\n",
    "        height=600\n",
    "    )\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fig_spotlen_bases(df, color_col):\n",
    "    fig = px.scatter(df, \n",
    "                     x=\"AvgSpotLen\", \n",
    "                     y=\"Bases\", \n",
    "                     color=color_col, \n",
    "                     log_x=True, \n",
    "                     log_y=True,\n",
    "                     hover_name=\"Run\",\n",
    "                     hover_data=df.columns,\n",
    "                     template=\"seaborn\",\n",
    "                     # marginal_x='',\n",
    "                     # marginal_y=''\n",
    "    )\n",
    "\n",
    "    fig.update_layout(\n",
    "        margin=dict(l=20, r=20, t=20, b=20),\n",
    "        height=600\n",
    "    )\n",
    "\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_stats(df):\n",
    "    sra_num = len(df)\n",
    "    biosample_num = len(df.BioSample.unique())\n",
    "    bioproj_num = len(df.BioProject.unique())\n",
    "\n",
    "    return f'{sra_num} Runs, {biosample_num} BioSamples, {bioproj_num} BioProjects'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bases_stats(df):\n",
    "    idx = df[\"AvgSpotLen\"]!='N/A'\n",
    "    avgspotlen = df.loc[idx, \"AvgSpotLen\"].mean()\n",
    "    \n",
    "    idx = df[\"Bases\"]!='N/A'\n",
    "    bases = df.loc[idx, \"Bases\"].mean()\n",
    "\n",
    "    return 'Mean AvgSpotLen: {:,.2f}bp, Mean Bases: {:,.2f}bp'.format(avgspotlen, bases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dropdown_div(dimensions_dict):\n",
    "    input_list = []\n",
    "\n",
    "    # add selection option for coloring column\n",
    "    options = [{'label': key, 'value': val} for key, val in dimensions_dict.items()]\n",
    "    div = html.Div([\n",
    "        html.Label('Coloring'),\n",
    "        dcc.Dropdown(\n",
    "            id='colored_column',\n",
    "            value='Assay Type',\n",
    "            options=options\n",
    "        ),\n",
    "    ], style={'width': '12%', 'display': 'inline-block'})\n",
    "\n",
    "    input_list.append(div)\n",
    "\n",
    "    # add selection options for dimensions_dict\n",
    "    for key in dimensions_dict:\n",
    "        dim = dimensions_dict[key]\n",
    "        available_idx = df[dim].unique()\n",
    "        options = [{'label': i, 'value': i} for i in available_idx]\n",
    "        div = html.Div([\n",
    "            html.Label(key),\n",
    "            dcc.Dropdown(\n",
    "                id=key,\n",
    "                options=options\n",
    "            ),\n",
    "        ], style={'width': '12%', 'display': 'inline-block'})\n",
    "\n",
    "        input_list.append(div)\n",
    "        \n",
    "    return input_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"650\"\n",
       "            src=\"http://127.0.0.1:8050/\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x1aa6e0cd0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "app = JupyterDash(__name__, external_stylesheets=[dbc.themes.BOOTSTRAP])\n",
    "\n",
    "input_list = dropdown_div(dimensions_dict)\n",
    "\n",
    "fig = fig_parallel_categories(df, dimensions_display, 'Assay Type')\n",
    "fig_sc = fig_spotlen_bases(df, 'Center Name')\n",
    "\n",
    "app.layout = html.Div([\n",
    "    html.H2(\n",
    "        children='Wastewater metagenome',\n",
    "        style={\n",
    "            'color': '#333333'\n",
    "        }\n",
    "    ),\n",
    "    html.Div(\n",
    "        id='stats_output',\n",
    "        children='# Runs, # BioSamples, # BioProjects', \n",
    "        style={\n",
    "            'color': '#777777'\n",
    "        }\n",
    "    ),\n",
    "    html.Div(\n",
    "        input_list, \n",
    "        style={'padding': '15px 5px'}\n",
    "    ),\n",
    "    html.Div([\n",
    "        dcc.Graph(\n",
    "            id='sra_sankey',\n",
    "            figure=fig\n",
    "        )\n",
    "    ], style={'width': '90%', 'display': 'inline-block', 'padding': '10px 20px'}),\n",
    "    html.Div(\n",
    "        id='bases_stats_output',\n",
    "        children='Mean AvgSpotLen: #, Mean Bases: #',\n",
    "        style={\n",
    "            'color': '#777777'\n",
    "        }\n",
    "    ),\n",
    "    html.Div([\n",
    "        dcc.Graph(\n",
    "            id='sra_scatter',\n",
    "            figure=fig_sc\n",
    "        )\n",
    "    ], style={'width': '90%', 'display': 'inline-block', 'padding': '10px 20px'}),\n",
    "    html.Div([\n",
    "        dbc.Button(\"Export SRA\", color=\"primary\", id=\"btn_csv\", className=\"mr-1\"),\n",
    "        dcc.Download(id=\"download-dataframe-csv\")\n",
    "    ]),\n",
    "], style={'margin': '10px 20px'})\n",
    "\n",
    "\n",
    "@app.callback(\n",
    "    Output('sra_sankey', 'figure'),\n",
    "    Output('sra_scatter', 'figure'),\n",
    "    Output('stats_output', 'children'),\n",
    "    Output('bases_stats_output', 'children'),\n",
    "    Input('assay_type', 'value'),\n",
    "    Input('library_source', 'value'),\n",
    "    Input('platform', 'value'),\n",
    "    Input('continent', 'value'),\n",
    "    Input('country', 'value'),\n",
    "    Input('colored_column', 'value'))\n",
    "def update_graph(assay_type, library_source, platform, continent, country, colored_column):\n",
    "    global ddf\n",
    "    ddf = df\n",
    "    if assay_type:\n",
    "        ddf = ddf[ddf[dimensions_dict['assay_type']]==assay_type]\n",
    "    if library_source:\n",
    "        ddf = ddf[ddf[dimensions_dict['library_source']]==library_source]\n",
    "    if platform:\n",
    "        ddf = ddf[ddf[dimensions_dict['platform']]==platform]\n",
    "    if continent:\n",
    "        ddf = ddf[ddf[dimensions_dict['continent']]==continent]\n",
    "    if country:\n",
    "        ddf = ddf[ddf[dimensions_dict['country']]==country]\n",
    "    if not colored_column:\n",
    "        colored_column = 'Assay Type'\n",
    "    \n",
    "    fig    = fig_parallel_categories(ddf, dimensions_display, colored_column)\n",
    "    fig_sc = fig_spotlen_bases(ddf, color_col='Center Name')\n",
    "    \n",
    "    return fig, fig_sc, get_stats(ddf), get_bases_stats(ddf)\n",
    "\n",
    "\n",
    "@app.callback(\n",
    "    Output(\"download-dataframe-csv\", \"data\"),\n",
    "    Input(\"btn_csv\", \"n_clicks\"),\n",
    "    prevent_initial_call=True,\n",
    ")\n",
    "def func(n_clicks):\n",
    "    global ddf\n",
    "    return dcc.send_data_frame(\n",
    "        df_sra[df_sra.Run.isin(ddf.Run)].to_csv,\n",
    "        \"sra_wastewater.csv\"\n",
    "    )\n",
    "\n",
    "# Run app and display result inline in the notebook\n",
    "app.run_server(mode='inline', debug=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
